### ISSUE TITLE
Add coursier to sbt under a flag

### ISSUE THREAD

**Issue by jvican** [sbt/sbt#2997]
The default ivy resolver in sbt is slow. Anyone that has started a sbt session in CI knows. The attempts that have been made to make it faster (for example, cached resolution in 0.13.7) still fall short. [Read this discussion for more details](https://github.com/sbt/sbt/issues/413).
  
Coursier is an sbt plugin that speeds up resolution times. [In one informal benchmark](https://github.com/sbt/sbt/pull/2992), Coursier resolves spark-sql in 13 seconds while the default resolver does it in 51 seconds.
  
I suggest that we satisfy the need of fast resolution times. I propose we merge a coursier-based artifact resolution. As sbt maintainers care about stability, I propose this is done under a flag. But in the future 0.13.x version we can consider removing the default resolver if no further issues with Coursier are encountered.
  
Coursier is almost ready to hit a 1.0 release. I propose that's the one we use. Some can say that Coursier resolution is not as good as ivy. I'd like to see some examples of this if it's true, but I'd like to point out that it's the other way around too. [This issue](https://github.com/sbt/sbt/issues/2953) shows an example of an artifact that is not resolved by Ivy but it is by Coursier.
  
One issue with adding Coursier is its external dependencies. We've been discussing in https://github.com/sbt/sbt/pull/2992 that there exists the possibility to include Coursier by shading them. To the best of my knowledge, @alexarchambault has offered his help to pull this effort off.
  
I have personally settled to parallelize ivy resolution times in this PR 
 https://github.com/sbt/sbt/pull/2992. I'd like this one to be merged just for those that are forced to use ivy resolution. However, I consider coursier to be a much better alternative and would like to see it as the default one, whether now or in the future.
  
A benefit of merging this into sbt over using it as a plugin is that it allows us to resolve sbt modules faster, which is important when installing sbt for the first time. If we're to recommend sbt as the default way of getting Scala, this operation should be fast.
  
Sbt users frequently complain about resolutions. Coursier is the best solution I know to this problem. 
  
I'd like to see whether sbt maintainers are willing to make this step. I'm happy to work alongside @alexarchambault to make this a reality.
  
Related discussion: https://github.com/sbt/sbt/issues/2294.

**Comment by fommil** [comment]
There are not enough thumbs on my hands to give this all the :+1: it needs

**Comment by dragos** [comment]
Is there a summary of the differences between the two resolvers? You imply that coursier may give different results, so it'd be good to understand these differences and how the two algorithms are different.

**Comment by DavidGregory084** [comment]
One thing I have noticed is that Coursier uses more memory than the default SBT resolver, so I hope that will be considered before it's proposed as the default resolver.

I haven't had much time to spend investigating this, but it reliably causes OOM on our Jenkins infrastructure where the default resolver does not (see also alexarchambault/coursier#399). Note this is for a large multi-module project so may not represent the average user of SBT very well.

**Comment by jastice** [comment]
Currently in the sbt integration of the IntelliJ Scala-Plugin, one of the biggest impacts on a smooth import experience is the ivy resolve times, as we need to run `update` to get all the transitive dependencies. I understand coursier reduces this problem quite a bit, and any change that cuts down on refresh times will be a big win for the user experience of plain sbt as well as in IDEA.

**Comment by dwijnand** [comment]
Hey @jvican thank you for taking the time to open and detail this issue, as well as bringing attention to it. It's definitely an important problem in sbt.

I'm all for this proposal. :+1:

---

Here are some notes about the topic:

> But in the future 0.13.x version we can consider removing [the opt-in flag] if no further issues with Coursier are encountered

Probably best to leave an opt-back flag, i.e to opt back to Ivy resolution.

> A benefit of merging this into sbt over using it as a plugin is that it allows us to resolve sbt modules faster, which is important when installing sbt for the first time. If we're to recommend sbt as the default way of getting Scala, this operation should be fast.

Agreed. Do know that this is tricky because that logic is in sbt/launcher not in this repo.

Speaking of which, here's a summary of when things get downloaded and by whom on launching sbt:

0. (optional) sbt-extras downloads the sbt/launcher jar `sbt-launch.jar`, using bash (the sbt/sbt-launcher-package scripts require a local jar instead)
1. sbt/launcher, which is a generic app launcher, downloads "sbt" and launches it, using Ivy
2. sbt downloads the user-global (~/.sbt/) and project-local (project/) plugins, using Ivy

Now once you've launched sbt:

* `update` will download your library dependencies (the binary jars), using Ivy
* `updateClassifiers` will download the sources and javadoc jars for your library dependencies, using Ivy
* `updateSbtClassifiers` will download sbt's sources and javadoc jars, using Ivy

**Comment by fommil** [comment]
*sigh* I really wish the sbt jar **was** the launcher (and a single jar at that), skipping step 0.

**Comment by eed3si9n** [comment]
See also related proposals:

- Split dependency resolution into resolve and downloading (lock dependency)  (https://github.com/sbt/sbt/issues/2989)
- Parallelize and cache download of Artifacts (https://github.com/sbt/sbt/issues/2982)

Also related that I would like to see in sbt 1.0, is HTTP client exposed to plugin authors as "Network API" (https://github.com/sbt/sbt/issues/2189). Gigahorse is my prototype project for that, Play-WS now has Play-free artifacts, and Coursier uses okhttp IIUC. If Coursier is included with its dependencies shaded, we might consider wrapping okhttp with Scala API.

**Comment by jvican** [comment]
> Also related that I would like to see in sbt 1.0
  
This proposal concerns only 0.13.x. It would be great if you guys forward port the implementation to 1.0!
  
@eed3si9n Will you accept a PR to add a coursier-based resolution engine then?

**Comment by alexandru** [comment]
Just dropping a note to say that I've been using Coursier and resolving dependencies is way faster than the default resolver.

It can't replace the default right now of course, it's still young and might have problems, but @alexarchambault has been very responsive to issues.

**Comment by dwijnand** [comment]
We shouldn't use the flag `--fast-resolution` though. Fast is relative - it's a bad name choice. (See "NIO", "NIO.2" ...)

**Comment by eed3si9n** [comment]
What's the upside of including this in sbt 0.13 behind the flag tho? Coursier already fairly popular, and there's a plugin to include it optionally. Is the purpose to speed up the plugins (meta build) resolution?

As I've written in https://github.com/sbt/sbt/issues/2294, for sbt 1.0 we should aim to abstract out the commonality between the implementations that's free of Ivy as librarymanagement API. As Dale implied there may be other implementations down the road.

**Comment by jvican** [comment]
> Fast is relative - it's a bad name choice. (See "NIO", "NIO.2" ...)
  
How fast resolution is can be measured objectively... though, and initial results show that Coursier is faster. That said, this is a small detail, I'm happy to reconsider the flag name, thanks for pointing it out!
  
Another alternative is to have `--resolution=coursier` or `--resolution=ivy`. Ivy resolution can be the default one for now.
  
> What's the upside of including this in sbt 0.13 behind the flag tho? Is the purpose to speed up the plugins (meta build) resolution?
  
Yes, the purposes are:
  
* Speed up sbt jars resolution for first installation (useful for beginners/us/CIs). The Scala community needs to improve the beginner experience.
* Provide built-in support for faster resolution using Coursier. Coursier is stable enough to deserve this and its speedups should be more accessible to the Scala community. If it proves to be better than ivy, then we can consider making it the default one.
  
There's no upside including this in sbt 0.13.x. If you want it to be swappable, I volunteer to provide an interface that both ivy and coursier resolve engines implement. Sbt 1.0 can provide fancier mechanisms to do so, but this is not the goal of this ticket.

**Comment by eed3si9n** [comment]
> - Speed up sbt jars resolution for first installation (useful for beginners/us/CIs). The Scala community needs to improve the beginner experience.

There's a PR that I'm working on (https://github.com/sbt/sbt/pull/2993) to include offline repository to sbt 0.13. If this goes through sbt will boot up in 10s the first time you run it.

```
$ time java -jar sbt-launch.jar -Dsbt.ivy.home=/tmp/offline/ivy.home -Dsbt.global.base=/tmp/offline/global exit
Getting org.scala-sbt sbt 0.13.14-SNAPSHOT ...
downloading file:/tmp/offline/global/preloaded/org.scala-sbt/sbt/0.13.14-SNAPSHOT/jars/sbt.jar ...
	[SUCCESSFUL ] org.scala-sbt#sbt;0.13.14-SNAPSHOT!sbt.jar (3ms)
downloading file:/tmp/offline/global/preloaded/org.scala-lang/scala-library/2.10.6/jars/scala-library.jar ...
	....
[info] Set current project to offline (in build file:/private/tmp/offline/)
java -jar sbt-launch.jar -Dsbt.ivy.home=/tmp/offline/ivy.home  exit  10.66s user 0.69s system 323% cpu 3.508 total
```

**Comment by cvogt** [comment]
@jvican we've been using Coursier at @xdotai since summer or so. Very early on there was one problem that alex fixed within days. Since then it's been smooth. There is [a several second overhead (link)](https://cloud.githubusercontent.com/assets/274947/23707370/15726820-03e0-11e7-8aed-eb7b3b8ee45c.png) (nothing compared to ivy). Not sure if that's Coursier, the sbt plugin or sbt itself.

I personally have no strong opinion on @jvican suggestion vs @eed3si9n question why not to just keep the plugin. The plugin works great for us and is automatically active for all of our developers.

On another note, CBT has it's own tiny resolver implementation. Yep, another one. It's not using Coursier or Ivy. CBT's resolver was designed for performance (no several second overhead), simple/tiny implementation, zero dependencies, not primarily completeness. It seems to work for many builds (e.g. our Spray/Play build at work), but CBT also plans to add Coursier support as a more complete alternative.

I am planning to pull CBT's resolver out into a self-contained library in the next few months. If any of this effort will make it easier to also allow using CBT's resolver with SBT, e.g. `--resolution=cbt` ;), I'd welcome this.

**Comment by alexarchambault** [comment]
@eed3si9n One would still have to download the content of the offline repo in a first time, isn't it? And what about bumping the version of sbt (that is, when running it with a new version in `project/build.properties`)?

**Comment by alexarchambault** [comment]
FYI, there was some discussion going on on the [Gitter of coursier](https://gitter.im/alexarchambault/coursier), about its custom sbt launcher in particular:
- the current master of coursier has a custom sbt launcher, with which all resolutions can be handled (checkout the [`demo-csbt` branch](https://github.com/alexarchambault/coursier/tree/demo-csbt) of coursier, optionally re-generate the launcher with `scripts/generate-sbt-launcher.sh`, and run `./csbt` - note that there are no `project/` on this branch, only a `sbt.properties`),
- but it has at least two limitations for now:
  - it's slower than e.g. sbt-extras, when the cache is full - but the resolution seems not to be the problem... it can possibly be some classloader setup issues,
  - when compiling several modules at the same time, I often ran into spurious compilation errors (e.g. errors in generated code for default values, in large automatic type class derivations, ...) - but similar errors were also seen without the custom sbt launcher, so the launcher may just be unearthing particular underlying issues...

**Comment by fommil** [comment]
I am personally not at all interested in sbt as a middleware provider (e.g. exposing http client). If I want an I/O abstraction, I'll use whatever lib I want. I hope such "sbt as a library" plans are not introducing unnecessary complexity.

**Comment by He-Pin** [comment]
In fact I think could we delegate the downloading to external tools as `axel` or `aria2c`?This will be a nice addition.
https://github.com/sbt/sbt/issues/1802

**Comment by eed3si9n** [comment]
@alexarchambault That's correct. The user story I intend to address the getting started experience, and to address no-internet installation in training or locked-down setups.

**Comment by eed3si9n** [comment]
@fommil If you don't want it you don't have to use it. There are a lot of plugins that require simple downloading of files, and the current situation is that every plugin introduces competing/conflicting wrapper of Netty/AHC or some other things.

**Comment by He-Pin** [comment]
I have not really looked much into the code,but without a proxy or coursier enabled ,there will be a lone time sbt show nothing on the console when you run `sbt ` in a new project.

We could simply shows how many artifacts need to resolving and what's going on now.for the non metadata downloading ,some of them are big ,and you could not easily download it with single thread downloading,So I want we could delegate the downloading to some external downloader.
If not,we could implement something like internetdownloadmanager or xunlei does in our internal downloader too.

In China,it's really not easy to make sbt resoving fast. 

Just my two cents input.

**Comment by fommil** [comment]
@eed3si9n my point is that resolution speed is a huge issue, whereas I don't hear anybody saying "gosh, I just wish sbt had a builtin io/network/serialisation library I could use" and it sounds like the latter is introducing complexity for the former.

If sbt didn't have to resolve itself, I suspect that many of these issues disappear (because meta resolving becomes a much smaller issue, especially on first startup, and the coursier plugin can be easily used for the main project)

**Comment by He-Pin** [comment]
@fommil Yes,enough extension point,for extension,if could not done it right and fast just get done.

**Comment by jvican** [comment]
> There's a PR that I'm working on (#2993) to include offline repository to sbt 0.13.
  
I'm glad somebody is finally working on this!
  
However, this doesn't weaken my point that the Scala community needs faster resolution times and  a Coursier resolver should be provided inside sbt and be pluggable. I don't see why we need to make sbt users pay the price of ivy resolution when there  are better alternatives.
  
@cvogt mentions another tool to resolve dependencies. What I propose allow us to easily support it by just adding a new flag that chooses that resolver. This is a flexible and generic solution that satisfies all the needs in the resolution front.
  
Coursier is not perfect in the same way ivy is not. Letting users select what they want is the most reasonable option.
  
Nevertheless, Coursier has an advantage over Ivy. It's a community project that we can improve and develop independently, which includes fixing misbehaviours. This advantage is huge and would also coordinate with the efforts of the sbt team to split sbt in different repositories.

**Comment by hedefalk** [comment]
> Some can say that Coursier resolution is not as good as ivy. I'd like to see some examples of this if it's true

Coursier is yet to handle moved artifacts which ivy2 does:

https://github.com/alexarchambault/coursier/issues/404

**Comment by fosskers** [comment]
I recently had to unrecommend coursier to the team I work on, because it started either:

- Blowing the heap trying to resolve deps
- Throwing NPEs at startup time (or during resolution)
- Failing to resolve deps for the [GeoWave project](https://github.com/ngageoint/geowave)

We work on a [large, complexly configured Geospatial computation library (GeoTrellis)](https://github.com/locationtech/geotrellis) which has been known to tax the capabilities of other Scala tools (i.e. Ensime), so I'm not entirely sure where to place "blame" here. Using the default ivy resolver takes half a century for GeoTrellis, so I'd love to see Coursier "blessed" as described in this issue.

Otherwise, Coursier works wonderfully with smaller projects.

**Comment by eed3si9n** [comment]
So maybe the middle ground is that we split resolution into two pieces. Dependency graph resolution and parallel caching/downloading. As I understand, the majority of the "win" from Coursier  comes from the latter part. Smart connection sharing of HTTPS and caching is something we can adopt relatively safely and turn it on by default.

**Comment by jvican** [comment]
> Coursier is yet to handle moved artifacts which ivy2 does
  
> unrecommend coursier to the team I work on
  
These issues are bugs, so if you file an issue in coursier, we'll try to get them fixed (or you can contribute a fix).
  
Adding a coursier-based resolution in sbt will also help us debug coursier's behaviour and spot these issues, making Coursier work for your project too.

**Comment by jvican** [comment]
> So maybe the middle ground is that we split resolution into two pieces. Dependency graph resolution and parallel caching/downloading. As I understand, the majority of the "win" from Coursier comes from the latter part.
  
@eed3si9n Actually, it's quite the contrary: coursier resolution is the one that beats up ivy's, not download times. Here's the informal benchmark results I use to compare, extracted from [this PR](https://github.com/sbt/sbt/pull/2992) and linked in the description of this ticket:

```
coursier: 13s
0.13.13: 51s
0.13.14-SNAPSHOT: 36s
```
  
As @olafurpg puts it:
> I noticed that resolution was taking the longest time in the plain sbt resolver projects. I re-ran the experiment for 0.13.13 on a new clean project to record how many seconds it took before downloading started. It took 32 seconds to resolve.
  
Ivy spends 32 seconds to *only* resolve (+ 19 seconds to download), while Coursier resolves **and** downloads in 13 seconds. This shows the potential of Coursier.

**Comment by non** [comment]
My 2¢: I think `--resolver=coursier` and `--resolver=ivy` (currently the default) is an elegant way to take advantage of Coursier when possible, without breaking existing builds. Ideally as a user of SBT I won't have to do more than add a flag or configuration option to explore Coursier's potential.

**Comment by fosskers** [comment]
I think coursier should be the default, provided it's kinks can be worked out.

**Comment by fommil** [comment]
@fosskers I use ensime on one of the largest scala codebases in the world (2+ million lines of scala, in a monorepo). You can either [contribute](http://ensime.org/contributing/) (preferred) or [sponsor](http://ensime.org/sponsor/) if you would like things to improve for your team. Unlike sbt, ensime is not backed by a commercial organisation and has much less stakeholders and [we've only heard about one thing from you before now, which was a bug in emacs](https://github.com/search?utf8=%E2%9C%93&q=user%3Aensime+geotrellis&type=Issues&ref=searchresults)

**Comment by fosskers** [comment]
Hey @fommil , how's it going. I was actually referring to something I mentioned in your Gitter many months ago, which was a spontaneous loss of semantic highlighting along with my general frequent need to restart Ensime to restore usual usage (of `C-c C-c e`, etc). At the time I think you mentioned that Ensime has trouble resolving complex implicit usage, which GeoTrellis has in spades. 

We can discuss this elsewhere if you're interested, since it's not directly related to this Coursier issue.

**Comment by fommil** [comment]
@fosskers +1 yeah let's relocate, although I don't have time to fix this myself (without a commercial reason) I can advise if you provide an example. Perhaps you could get involved with the proposed workaround, it's low on my list but almost done. Frankly, I could fix it with a couple of days.

**Comment by cvogt** [comment]
@eed3si9n I share @jvican's impression that resolution speed is the problem with ivy/sbt, not downloading speed. I am usually on reliable, high speed connections though, so this may depend on location.

**Comment by eed3si9n** [comment]
It really depends on the size of the graph and network latency etc, but for example I've asked @gakuzzzz to benchmark when he blogged about sbt-coursier, and for his project the first run and second run was:

- 1238s / 27s with Ivy
- 121s / 13s with sbt-coursier

https://gist.github.com/gakuzzzz/49b78acfe1401fd8047c7a0f0d981cfa

https://github.com/sbt/sbt/issues/2189 also explains how Ivy doesn't share HTTPS connections, resulting in paying latency penalties many times.

**Comment by g-dumas** [comment]
++++++100000000

**Comment by scottcarey** [comment]
Everyone is mentioning speed.  Yes it is faster.  Yes, that is great.  But for me, my build WILL NOT WORK at all with the default resolver in some cases.  

My situation:  I depend on an artifact that is built via maven and found only in a corporate repo.   If I build a snapshot of that maven artifact locally and publish to my local maven repo, the default resolver gets completely confused and mixes up the ivy local and maven local, often being unable to locate the artifact.   I tried every suggestion and fix I could find on Stack Overflow and in the documentation.  Nothing worked.  I turned on courseir, and everything went smoothly.   My setup includes a couple quirks, like _disabling_ maven central and forcing everything through a corporate proxy.  The default resolver just couldn't handle it.

Coursier has much better support for maven repos and local maven repositories, in addition to all the ivy stuff.

+∞
